contracts/examples/bonding-curve/src/utils/owner_endpoints.rs

```diff
@@ -0,0 +1,169 @@
+	}
+
+	#[endpoint(unsetLocalRoles)]
+	fn unset_local_roles(
+		&self,
+		address: Address,
+		token_identifier: TokenIdentifier,
+		#[var_args] roles: VarArgs<EsdtLocalRole>,
+	) -> AsyncCall<Self::SendApi> {
+		ESDTSystemSmartContractProxy::new_proxy_obj(self.send())
+			.unset_special_roles(&address, &token_identifier, roles.as_slice())
+			.async_call()
+			.with_callback(OwnerEndpointsModule::callbacks(self).change_roles_callback())
+	}
+
+	#[callback]
+	fn change_roles_callback(&self, #[call_result] result: AsyncCallResult<()>) -> SCResult<()> {
+		match result {
+			AsyncCallResult::Ok(()) => Ok(()),
+			AsyncCallResult::Err(message) => Err(message.err_msg.into()),
+		}
+	}
+
+	#[endpoint(setBondingCurve)]
+	fn set_bonding_curve(
+		&self,
+		identifier: TokenIdentifier,
+		function: FunctionSelector<Self::BigUint>,
+	) -> SCResult<()> {
+		require!(
+			!self.token_details(&identifier).is_empty(),
+			"Token is not issued yet!"
+		);
+
+		let caller = self.blockchain().get_caller();
+
+		let details = self.token_details(&identifier).get();
+		require!(
+			details.owner == caller,
+			"The price function can only be set by the seller."
+		);
+		self.bonding_curve(&identifier)
+			.update(|bonding_curve| bonding_curve.curve = function);
+		Ok(())
+	}
+
+	#[endpoint(deposit)]
+	#[payable("*")]
+	fn deposit(
+		&self,
+		#[payment] amount: Self::BigUint,
+		#[payment_token] identifier: TokenIdentifier,
+		#[payment_nonce] nonce: u64,
+		#[var_args] payment_token: OptionalArg<TokenIdentifier>,
+	) -> SCResult<()> {
+		let caller = self.blockchain().get_caller();
+		let mut set_payment = TokenIdentifier::egld();
+		if self.bonding_curve(&identifier).is_empty() {
+			set_payment = payment_token
+				.into_option()
+				.ok_or("Expected provided accepted_payment for the token")?;
+		}
+		if self.token_details(&identifier).is_empty() {
+			self.token_details(&identifier).set(&TokenOwnershipData {
+				token_nonces: [nonce].to_vec(),
+				owner: caller.clone(),
+			});
+		} else {
+			let details = self.token_details(&identifier).get();
+			require!(
+				details.owner == caller,
+				"The token was already deposited by another address"
+			);
+			if !details.token_nonces.contains(&nonce) {
+				self.token_details(&identifier)
+					.update(|new_details| new_details.token_nonces.push(nonce));
+			}
+		}
+
+		self.set_curve_storage(&identifier, amount.clone(), set_payment);
+		self.owned_tokens(&caller).insert(identifier.clone());
+		self.nonce_amount(&identifier, nonce)
+			.update(|current_amount| *current_amount += amount);
+		Ok(())
+	}
+
+	#[endpoint(claim)]
+	fn claim(&self) -> SCResult<()> {
+		let caller = self.blockchain().get_caller();
+		require!(
+			!self.owned_tokens(&caller).is_empty(),
+			"You have nothing to claim"
+		);
+		for token in self.owned_tokens(&caller).iter() {
+			let nonces = self.token_details(&token).get().token_nonces;
+			for nonce in nonces {
+				self.send().direct(
```

Comment:
Just making sure, there is no way any of the `self.nonce_amount(&token, nonce).get()` are zero, right? Asking because trying to transfer zero would crash.

Reply:
nope.. you cannot pass the check with a value of 0